cpp_exercises/day2/date/date.cpp

```diff
@@ -137,3 +138,43 @@ void Date::fromJulianDay(int64_t julianDay) {
 void Date::addDays(int days) {
     fromJulianDay(toJulianDay() + days);
 }
+
+Date Date::fromString(const std::string& string) {
+    if(std::count(string.begin(), string.end(), ' ') != 0)
+        return Date();
+
+    if(std::count(string.begin(), string.end(), '-') != 2)
+        return Date();
+
+    std::string dayString, monthString, yearString;
+    Date::splitDayMonthYearString(string, dayString, monthString, yearString);
+
+    if(dayString.size() != 2 || monthString.size() != 2 || yearString.size() != 4)
+        return Date();
+
+    if(!Date::areStringValidInteger(dayString) || !Date::areStringValidInteger(monthString) || !Date::areStringValidInteger(yearString))
+        return Date();
+
+    int day = std::stoi(dayString);
+    int month = std::stoi(monthString);
+    int year = std::stoi(yearString);
+
+    if(!Date::isDateValid(day, month, year))
+        return Date();
+
+    return Date(day, month, year);
+}
+
+bool Date::areStringValidInteger(const std::string& string) {
+    char* p;
+    strtol(string.c_str(), &p, 10);
+    return *p == 0;
+}
+
+void Date::splitDayMonthYearString(const std::string& string, std::string& dayString, std::string& monthString, std::string& yearString) {
+    char separator = '-';
+    std::stringstream stringStream(string);
+    std::getline(stringStream, dayString, separator);
+    std::getline(stringStream, monthString, separator);
+    std::getline(stringStream, yearString, separator);
```

Comment:
Secondo me è un po' eccessivo l'uso di stringstream: nel nostro caso sappiamo benissimo che il formato della data deve essere `dd-mm-yyyy`, perciò secondo me tanto vale costruire delle stringhe usando il metodo `std::string::substr(...)`.
Possiamo aggiungere un paio di assert che controllino che ci siano i due separatori `-` al posto giusto volendo, ma in teoria a questo punto qualunque tipo di separatore andrebbe bene.